File: src/main/java/com/example/cqrs/repository/write/AccountEventRepository.java
package com.example.cqrs.repository.write;

import com.example.cqrs.entity.write.event.base.BaseAccountEvent;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;

/**
 * writeEntityManagerFactory와 writeTransactionManager 사용
 * @Primary로 지정되어 있어 별도 설정 없이도 쓰기 DB 사용
 */
@Repository
public interface AccountEventRepository extends JpaRepository<BaseAccountEvent, Long> {

    List<BaseAccountEvent> findByAccountIdAndEventDateAfterOrderByEventDateAsc(
            String accountId, LocalDateTime afterDate);

    List<BaseAccountEvent> findByAccountIdOrderByEventDateDesc(String accountId);

    List<BaseAccountEvent> findByAccountIdOrderByEventDateAsc(String accountId);

    List<BaseAccountEvent> findByMetadataCorrelationId(String correlationId);

    List<BaseAccountEvent> findByMetadataUserIdOrderByEventDateDesc(String userId);

    long countByAccountIdAndEventDateAfter(String accountId, LocalDateTime afterDate);

}

File: src/main/java/com/example/cqrs/command/repository/AccountEventRepository.kt
package com.example.cqrs.command.repository

import com.example.cqrs.command.entity.event.base.AccountEvent
import org.springframework.data.jpa.repository.JpaRepository
import java.time.LocalDateTime

/**
 * 계좌 이벤트 리포지토리
 * 모든 계좌 관련 이벤트 저장 및 조회
 */
interface AccountEventRepository : JpaRepository<AccountEvent, Long> {
    // 계좌 ID와 날짜 기준 이벤트 조회 (시간순 정렬)
    fun findByAccountIdAndEventDateAfterOrderByEventDateAsc(
        accountId: String, afterDate: LocalDateTime
    ): List<AccountEvent>

    // 계좌별 전체 이벤트 조회 (최신순 정렬)
    fun findByAccountIdOrderByEventDateDesc(accountId: String): List<AccountEvent>

    // 계좌별 전체 이벤트 조회 (시간순 정렬)
    fun findByAccountIdOrderByEventDateAsc(accountId: String): List<AccountEvent>

    // 연관 ID로 이벤트 그룹 조회
    fun findByMetadataCorrelationId(correlationId: String): List<AccountEvent>

    // 사용자별 이벤트 조회 (최신순 정렬)
    fun findByMetadataUserIdOrderByEventDateDesc(userId: String): List<AccountEvent>

    // 특정 날짜 이후 이벤트 개수 조회
    fun countByAccountIdAndEventDateAfter(accountId: String, afterDate: LocalDateTime): Long
}